File: backend/app/schemas/paper.py
import datetime

from pydantic import BaseModel, Field

from app.schemas.common import PaginatedResponse


class PaperFilterParams(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    sort_by: str = Field(default="publication_date")
    sort_order: str = Field(default="desc")
    sci_zone: list[str] | None = Field(default=None, alias="sci_zone[]")
    source: list[str] | None = Field(default=None, alias="source[]")
    keyword_id: list[int] | None = Field(default=None, alias="keyword_id[]")
    date_from: str | None = None
    date_to: str | None = None
    citations_min: int | None = None
    q: str | None = None
    has_summary: bool | None = None
    starred: bool | None = None

    model_config = {"extra": "allow"}


class PaperResponse(BaseModel):
    id: int
    title: str
    authors: str
    abstract: str | None
    publication_date: str | None
    source: str
    doi: str | None
    arxiv_id: str | None
    url: str | None
    pdf_url: str | None
    journal_name: str | None
    sci_zone: str | None
    citation_count: int
    year: int | None
    is_starred: bool
    has_summary: bool = False
    summary_status: str = "none"  # "completed" | "processing" | "none"
    keyword_texts: list[str] = []
    keyword_ids: list[int] = []
    crawled_at: datetime.datetime | None = None
    updated_at: datetime.datetime | None = None

    model_config = {"from_attributes": True}


class PaperListResponse(PaginatedResponse):
    papers: list[PaperResponse]


class PaperDetailResponse(PaperResponse):
    model_config = {"from_attributes": True, "protected_namespaces": ()}

    summary_cn: str | None = None
    key_points_cn: str | None = None
    model_used: str | None = None
    summary_generated_at: datetime.datetime | None = None
    source_type: str | None = None
    source_chars: int = 0


class PaperStarRequest(BaseModel):
    starred: bool


class PaperStatsResponse(BaseModel):
    total: int
    with_summary: int
    by_zone: dict[str, int]
    by_source: dict[str, int]

File: frontend/src/types/paper.ts
export interface Paper {
  id: number;
  title: string;
  authors: string;
  abstract: string | null;
  publication_date: string | null;
  source: string;
  doi: string | null;
  arxiv_id: string | null;
  url: string | null;
  pdf_url: string | null;
  journal_name: string | null;
  sci_zone: string | null;
  citation_count: number;
  year: number | null;
  is_starred: boolean;
  has_summary: boolean;
  summary_status: 'completed' | 'processing' | 'none';
  keyword_texts: string[];
  keyword_ids: number[];
  crawled_at: string | null;
  updated_at: string | null;
}

export interface PaperDetail extends Paper {
  summary_cn: string | null;
  key_points_cn: string | null;
  model_used: string | null;
  summary_generated_at: string | null;
  source_type: string | null;
  source_chars: number;
}

export interface PaperListResponse {
  total: number;
  page: number;
  page_size: number;
  papers: Paper[];
}

export interface PaperStats {
  total: number;
  with_summary: number;
  by_zone: Record<string, number>;
  by_source: Record<string, number>;
}

export interface PaperFilterParams {
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: string;
  'sci_zone[]'?: string[];
  'source[]'?: string[];
  'keyword_id[]'?: number[];
  date_from?: string;
  date_to?: string;
  citations_min?: number;
  q?: string;
  has_summary?: boolean;
  starred?: boolean;
}
